docs/developer-docs/smart-contracts/maintain/settings.mdx

```diff
@@ -117,3 +117,107 @@ async fn modify_canister_settings() -> CallResult<()> {
+
+## Reserved cycles limit
+
+In addition to the main cycles balance, a canister has a secondary cycles balance called reserved cycles.
+The reserved cycles are cycles that are set aside for future resource payments by the resource reservation mechanism.
+You can learn more about resource reservation [here](docs/current/developer-docs/gas-cost).
+
+The `reserved_cycles_limit` field in canister settings serves as the upper limit on reserved cycles.
+It protects against spending too many cycles in resource reservation.
+
+The default value is 5 trillion cycles.
+
+## Wasm memory limit
+
+A canister has a Wasm memory and a stable memory.
+Currently, the Wasm memory is 32-bit, which means that it cannot grow beyond the hard limit of 4GiB.
+
+If a canister stores user data in the Wasm memory instead of the stable memory, then its memory usage will increase with the number of users.
+Eventually, such a canister could reach 4GiB at which point it won’t be able to allocate more memory and will stop working.
+It is also likely that the pre-upgrade hook will fail to allocate, which will make the canister and its data unrecoverable.
+Even if the canister doesn’t store user data in Wasm memory, its memory usage may increase due to memory leaks.
+
+The `wasm_memory_limit` field in canister settings provides "best-effort" protection against canister running out of Wasm memory.
+The limit is expressed in bytes and it is a soft limit, which means that in some circumstances the canister may grow its Wasm memory above the limit.
+Handling of the limit depends on the message type:
+
+- **update messages**: the limit is enforced up to the first `await` point.
+If the canister attempts to grow its Wasm memory above the limit, then execution fails with the corresponding error message.
+After the `await` point, execution continues in a response callback where the limit is not enforced.
+- **response callback**: the limit is not enforced and the canister may grow its Wasm memory above the limit.
+The motivation for not enforcing the limit is to avoid failing execution after the `await` point that could lead to subtle bugs in the canister code.
+- **hearbeats and timers**: the limit is currently not enforced, but will be enforced in future after the canister logging feature is available.
+Without canister logging it is difficult for developers to debug failures in heartbeat and timers.
+- **queries**: the limit is not enforced in queries because state changes are not preserved in query execution so any increase in memory is tentative and will be reverted after the query execution completes.
+- **canister init**: the limit is enforced and code installation fails if it grows the Wasm memory above the limit.
+- **canister pre-upgrade**: the limit is not enforced in order to allow upgrading the canister to a new version.
+- **canister post-upgrade**: the limit is enforced and code upgrade fails if the new version has Wasm memory usage higher than the limit.
```

Comment:
```suggestion
- **Update messages**: Enforced up to the first `await` point. If the canister attempts to grow its Wasm memory above the limit, then execution fails with a corresponding error message. After the `await` point, execution continues in a response callback where the limit is not enforced.
- **Response callback**: Not enforced and the canister may grow its Wasm memory above the limit. The motivation for not enforcing the limit is to avoid failing execution after the `await` point that could lead to subtle bugs in the canister code.
- **Heartbeats and timers**: Currently not enforced, but will be enforced in the future after the canister logging feature is available. Without canister logging it is difficult for developers to debug failures in heartbeats and timers.
- **Queries**: Not enforced in queries because state changes are not preserved in query execution, so any increase in memory is tentative and will be reverted after the query execution completes.
- **Canister init**: Enforced and code installation fails if it grows the Wasm memory above the limit.
- **Canister pre-upgrade**: Not enforced in order to allow upgrading the canister to a new version.
- **Canister post-upgrade**: Enforced and code upgrades will fail if the new version has Wasm memory usage higher than the limit.
```